File: runner/mission_runner/preview.py
"""Two ways to see what a mission will do before the robot moves.

* :func:`plan_route` asks the real planner for the path between every
  navigation step, so the editor can draw the route the robot would actually
  drive instead of straight lines, with distance and a time estimate.
* :class:`DryRun` replays the whole flow - branches, loops, retries and all -
  against a private simulated robot seeded at the current pose, and returns a
  timeline the editor can animate. Nothing reaches the real robot, the
  connectors or the run log.
"""

from __future__ import annotations

import asyncio
import logging
import math
import time
from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Any

from .backends.base import Pose, path_length
from .backends.sim import SimBackend
from .connectors.base import Armed, Connector, ConnectorRegistry
from .events import EventBus
from .model import EventSource, Mission, Step
from .types import Run, RunSource, RunStatus, StepFailed

if TYPE_CHECKING:
    from .runner import Runner

log = logging.getLogger("mission.preview")

#: Steps that move the robot along a route, in the order the editor draws them.
ROUTE_STEPS = ("nav.go_to_pose", "nav.go_through_poses", "nav.follow_waypoints", "nav.follow_path")
#: Rough fixed costs, seconds, for steps that take time but cover no distance.
FIXED_COST_S = {"nav.dock": 25.0, "nav.undock": 8.0, "nav.spin": 4.0, "nav.change_map": 3.0, "nav.clear_costmap": 1.0, "ask_user": 0.0}


# ---------------------------------------------------------------------------
# route preview


@dataclass
class Leg:
    """One planned move between two poses."""

    step_id: str
    step_type: str
    index: int
    start: dict[str, Any]
    goal: dict[str, Any]
    path: dict[str, Any] | None = None
    length_m: float = 0.0
    planned: bool = False
    error: str = ""

    def as_dict(self) -> dict[str, Any]:
        return {
            "step_id": self.step_id,
            "step_type": self.step_type,
            "index": self.index,
            "start": self.start,
            "goal": self.goal,
            "path": self.path,
            "length_m": round(self.length_m, 2),
            "planned": self.planned,
            "error": self.error,
        }


def static_route(mission: Mission, resolve: Any) -> list[tuple[Step, int, Pose]]:
    """Walk the flow the way the editor draws it: navigation steps in order,
    entering `then` branches and loop bodies once. Returns (step, index, pose)."""
    out: list[tuple[Step, int, Pose]] = []

    def walk(steps: list[Step]) -> None:
        for s in steps:
            if not s.enabled:
                continue
            if s.type == "if":
                walk(s.container("then"))
                continue
            if s.type == "loop":
                walk(s.container("body"))
                continue
            if s.type not in ROUTE_STEPS:
                continue
            for i, pose in enumerate(_poses_of(s, resolve)):
                out.append((s, i, pose))

    walk(mission.flow)
    return out


def _poses_of(step: Step, resolve: Any) -> list[Pose]:
    key = {"nav.go_to_pose": "pose", "nav.go_through_poses": "poses", "nav.follow_waypoints": "poses", "nav.follow_path": "points"}[step.type]
    raw = step.params.get(key)
    if raw is None:
        return []
    values = raw if isinstance(raw, list) else [raw]
    poses: list[Pose] = []
    for v in values:
        try:
            poses.append(resolve(v))
        except Exception:  # noqa: BLE001 - an expression we cannot resolve statically
            continue
    return poses


async def plan_route(runner: Runner, mission: Mission, start: Pose | None = None, planner_id: str = "") -> dict[str, Any]:
    """Plan every leg of ``mission`` with the real planner. Falls back to
    straight lines when the planner is unavailable or the robot is busy."""

    scope = _static_scope(runner, mission)

    def resolve(raw: Any) -> Pose:
        return runner.interpreter.resolve_pose(raw, scope)

    stops = static_route(mission, resolve)
    rs = runner.backend.robot_state()
    cursor = start or (Pose(rs.x, rs.y, rs.yaw_deg or 0.0) if rs.x is not None and rs.y is not None else None)
    busy = runner.dispatcher.busy()
    legs: list[Leg] = []
    for n, (step, idx, goal) in enumerate(stops):
        leg = Leg(step.id, step.type, idx, cursor.as_dict() if cursor else {}, goal.as_dict())
        if cursor is not None:
            if busy:
                leg.error = "robot is busy; showing a straight line"
            else:
                try:
                    leg.path = await asyncio.wait_for(runner.backend.compute_path(cursor, goal, planner_id, True), timeout=10)
                    leg.length_m = path_length(leg.path)
                    leg.planned = True
                except (StepFailed, asyncio.TimeoutError) as e:
                    leg.error = str(e) or "planner did not answer"
                except Exception as e:  # noqa: BLE001
                    leg.error = f"{type(e).__name__}: {e}"
            if not leg.planned:
                leg.length_m = cursor.distance_to(goal)
                leg.path = {"frame": goal.frame, "poses": [cursor.as_dict(), goal.as_dict()]}
        cursor = goal
        legs.append(leg)
        _ = n

    distance = sum(x.length_m for x in legs)
    speed = _nominal_speed(runner)
    fixed = sum(FIXED_COST_S.get(s.type, 0.0) for s in mission.iter_steps(mission.flow) if s.enabled)
    waits = sum(_wait_seconds(s) for s in mission.iter_steps(mission.flow) if s.enabled)
    return {
        "mission": mission.name,
        "legs": [x.as_dict() for x in legs],
        "distance_m": round(distance, 2),
        "estimate_s": round(distance / max(0.05, speed) + fixed + waits, 1),
        "speed_mps": speed,
        "planned": all(x.planned for x in legs) and bool(legs),
        "planner": runner.backend.name,
        "note": "robot is busy; distances are straight-line" if busy else "",
    }


def _wait_seconds(step: Step) -> float:
    if step.type != "wait":
        return 0.0
    v = step.params.get("seconds")
    return float(v) if isinstance(v, (int, float)) else 0.0


def _nominal_speed(runner: Runner) -> float:
    b = runner.backend
    return float(getattr(b, "speed", 0.0)) or float((runner.config.sim or {}).get("speed_mps", 0.5)) or 0.5


def _static_scope(runner: Runner, mission: Mission) -> dict[str, Any]:
    """Enough scope to resolve poses that do not depend on run-time values."""
    scope: dict[str, Any] = {name: inp.default for name, inp in mission.inputs.items()}
    scope.update(mission.vars)
    scope["robot"] = runner.robot_scope()
    scope["current_map"] = runner.current_map
    return scope


# ---------------------------------------------------------------------------
# dry run


class _StubConnector(Connector):
    """Accepts everything a mission sends and records it."""

    type = "dryrun"
    source_types = ("mqtt.subscribe", "modbus.poll", "ros.topic", "gpio.input", "http.webhook", "timer.cron", "timer.interval", "timer.boot", "mission.done")
    singleton = True

    def __init__(self, sink: list[dict[str, Any]]):
        super().__init__("dryrun", {})
        self.sink = sink

    async def arm(self, source: EventSource, callback: Any) -> Armed:
        return Armed(lambda: None, source.summary())

    async def publish(self, topic: str, payload: Any, qos: int = 1, retain: bool = False) -> None:
        self.sink.append({"kind": "mqtt.publish", "topic": topic, "payload": payload})

    async def write(self, kind: str, address: int, value: Any) -> None:
        self.sink.append({"kind": f"{kind}.write", "address": address, "value": value})


class _StubRegistry(ConnectorRegistry):
    def __init__(self, stub: _StubConnector):
        super().__init__()
        self._stub = stub
        self.add(stub)

    def require(self, name: str) -> Connector:
        return self._stub

    def for_source(self, source: EventSource) -> Connector:
        return self._stub


class _StubTriggers:
    """`wait_event` is assumed to fire straight away, so a preview never hangs."""

    def __init__(self, notes: list[str]):
        self.notes = notes

    async def wait_event(self, source: EventSource, timeout_s: float | None) -> dict[str, Any]:
        self.notes.append(f"assumed '{source.summary()}' fires immediately")
        await asyncio.sleep(0)
        return {"dry_run": True}


class _StubPrompts:
    """`ask_user` answers with its default, or the first option."""

    def __init__(self, notes: list[str]):
        self.notes = notes
        self.current = None

    async def ask(self, *, run_id: str, mission: str, step_id: str, text: str, options: list[str], default: str | None, timeout_s: float | None) -> str:
        answer = default or (options[0] if options else "")
        self.notes.append(f"answered '{text}' with '{answer}'")
        await asyncio.sleep(0)
        return answer


@dataclass
class DryRunResult:
    ok: bool = True
    status: str = "succeeded"
    error: str = ""
    duration_s: float = 0.0
    distance_m: float = 0.0
    samples: list[dict[str, Any]] = field(default_factory=list)
    steps: list[dict[str, Any]] = field(default_factory=list)
    notes: list[str] = field(default_factory=list)
    outputs: list[dict[str, Any]] = field(default_factory=list)
    truncated: bool = False

    def as_dict(self) -> dict[str, Any]:
        return {
            "ok": self.ok,
            "status": self.status,
            "error": self.error,
            "duration_s": round(self.duration_s, 1),
            "distance_m": round(self.distance_m, 2),
            "samples": self.samples,
            "steps": self.steps,
            "notes": self.notes,
            "outputs": self.outputs,
            "truncated": self.truncated,
        }


class DryRun:
    """Runs a mission against a private simulated robot and records a timeline."""

    #: simulated seconds between timeline samples
    SAMPLE_S = 0.25

    def __init__(self, runner: Runner, mission: Mission, inputs: dict[str, Any] | None = None, *, time_scale: float = 60.0, max_wall_s: float = 20.0, max_sim_s: float = 3600.0):
        self.runner = runner
        self.mission = mission
        self.inputs = dict(inputs or {})
        self.time_scale = time_scale
        self.max_wall_s = max_wall_s
        self.max_sim_s = max_sim_s
        self.notes: list[str] = []
        self.outputs: list[dict[str, Any]] = []
        self.events = EventBus()
        rs = runner.backend.robot_state()
        start = Pose(rs.x or 0.0, rs.y or 0.0, rs.yaw_deg or 0.0)
        self.backend = SimBackend(speed_mps=_nominal_speed(runner), time_scale=time_scale, start=start, battery=rs.battery if rs.battery is not None else 0.85)
        self.registry = _StubRegistry(_StubConnector(self.outputs))
        self.triggers = _StubTriggers(self.notes)
        self.prompts = _StubPrompts(self.notes)
        self.store = runner.store
        self._map = runner.current_map
        self._step: dict[str, Any] | None = None
        self._samples: list[dict[str, Any]] = []
        self._steps: list[dict[str, Any]] = []
        self._t0 = 0.0

    # ----- Services -------------------------------------------------------------

    @property
    def current_map(self) -> str | None:
        return self._map

    def set_current_map(self, name: str | None) -> None:
        self._map = name

    def robot_scope(self) -> dict[str, Any]:
        return self.backend.robot_state().as_dict() or {}

    async def run_sub_mission(self, ctx: Any, name: str, inputs: dict[str, Any]) -> Any:
        from .types import StepResult

        sub = self.store.get(name)
        if sub is None:
            raise StepFailed(f"unknown mission '{name}'")
        self.notes.append(f"ran sub-mission '{name}'")
        run = Run(id=f"dry-{name}", mission=name, inputs=dict(inputs), source=RunSource("sub", "dryrun"))
        sub_ctx = self.interpreter.make_context(run, sub)
        sub_ctx.depth = getattr(ctx, "depth", 0) + 1
        outcome = await self.interpreter.run(sub_ctx)
        return StepResult(outcome.status == RunStatus.SUCCEEDED, outcome.status.value, outcome.result, outcome.error)

    async def http_request(self, method: str, url: str, headers: dict[str, str] | None, body: Any, timeout_s: float | None) -> dict[str, Any]:
        self.outputs.append({"kind": "http.request", "method": method, "url": url})
        self.notes.append(f"skipped {method} {url}")
        return {"status": 200, "body": {"dry_run": True}}

    # ----- execution ----------------------------------------------------------------

    async def execute(self) -> DryRunResult:
        from .interpreter import Interpreter

        self.interpreter = Interpreter(self)  # type: ignore[arg-type]
        self.events.subscribe(self._on_event)
        run = Run(id="dry-run", mission=self.mission.name, inputs=self.inputs, source=RunSource("manual", "dryrun"))
        try:
            ctx = self.interpreter.make_context(run, self.mission)
        except StepFailed as e:
            return DryRunResult(False, "failed", str(e), notes=self.notes)

        result = DryRunResult(notes=self.notes, outputs=self.outputs)
        self._t0 = time.monotonic()
        sampler = asyncio.create_task(self._sample_loop(), name="dryrun-sampler")
        try:
            outcome = await asyncio.wait_for(self.interpreter.run(ctx), timeout=self.max_wall_s)
            result.status = outcome.status.value
            result.ok = outcome.status == RunStatus.SUCCEEDED
            result.error = outcome.error
        except asyncio.TimeoutError:
            result.status = "timeout"
            result.ok = False
            result.truncated = True
            result.error = f"preview stopped after {self.max_wall_s:g} s of wall clock ({self.time_scale:g}x)"
            await self.backend.cancel()
        except Exception as e:  # noqa: BLE001
            result.status = "failed"
            result.ok = False
            result.error = f"{type(e).__name__}: {e}"
        finally:
            sampler.cancel()
            self._snapshot()
        result.samples = self._samples
        result.steps = self._steps
        result.duration_s = (time.monotonic() - self._t0) * self.time_scale
        result.distance_m = self.backend.odometer
        return result

    async def _sample_loop(self) -> None:
        interval = self.SAMPLE_S / self.time_scale
        while True:
            self._snapshot()
            await asyncio.sleep(max(0.005, interval))

    def _snapshot(self) -> None:
        if len(self._samples) >= 4000:
            return
        st = self.backend.robot_state()
        t = (time.monotonic() - self._t0) * self.time_scale
        last = self._samples[-1] if self._samples else None
        x, y, yaw = round(st.x or 0.0, 3), round(st.y or 0.0, 3), round(st.yaw_deg or 0.0, 1)
        step_id = (self._step or {}).get("id", "")
        if last and last["x"] == x and last["y"] == y and last["yaw_deg"] == yaw and last["step"] == step_id:
            return
        self._samples.append({"t": round(t, 2), "x": x, "y": y, "yaw_deg": yaw, "step": step_id})

    def _on_event(self, event: dict[str, Any]) -> None:
        t = round((time.monotonic() - self._t0) * self.time_scale, 2)
        kind = event.get("type", "")
        if kind == "step.started":
            self._step = {"id": event["step_id"], "name": event.get("name", ""), "type": event.get("step_type", ""), "t0": t, "path": event.get("path")}
            self._snapshot()
        elif kind == "step.finished":
            self._snapshot()
            cur = self._step or {"id": event["step_id"]}
            res = event.get("result") or {}
            self._steps.append({**cur, "t1": t, "status": res.get("status", ""), "error": res.get("error", ""), "duration_s": res.get("duration_s", 0)})
            self._step = None
        elif kind == "log":
            self.notes.append(f"{event.get('level', 'info')}: {event.get('text', '')}")

    # `time_scale` is a plain attribute set in __init__; the interpreter reads it
    # off the services object to compress `wait` steps by the same factor.

File: editor/src/map/preview.ts
/**
 * The two previews the runner can compute without moving the robot:
 * `POST /api/preview/route` (what the real planner would drive) and
 * `POST /api/preview/dryrun` (the whole flow replayed against a private
 * simulated robot). This module only parses and formats — no DOM, no fetch —
 * so the map view and the transport bar share one model.
 */

import type { WorldPoint } from "./geometry";
import { isRecord } from "../model/types";

const num = (v: unknown, dflt = 0): number => (typeof v === "number" && Number.isFinite(v) ? v : dflt);
const str = (v: unknown): string => (typeof v === "string" ? v : "");

// ---- route preview ------------------------------------------------------------------

export interface PreviewLeg {
  /** Stable key for the SVG node. */
  key: string;
  stepId: string;
  index: number;
  /** The planned polyline (at least the two endpoints). */
  points: WorldPoint[];
  planned: boolean;
  error: string;
  lengthM: number;
}

export interface RoutePreview {
  legs: PreviewLeg[];
  distanceM: number;
  estimateS: number;
  speedMps: number;
  /** Every leg came from the real planner. */
  planned: boolean;
  /** "robot is busy; distances are straight-line" and friends. */
  note: string;
  /** Document version this was computed for (cleared when the mission changes). */
  version: number;
  at: number;
}

function posePoints(raw: unknown): WorldPoint[] {
  const poses = isRecord(raw) && Array.isArray(raw.poses) ? raw.poses : [];
  const out: WorldPoint[] = [];
  for (const p of poses) {
    if (!isRecord(p)) continue;
    if (typeof p.x !== "number" || typeof p.y !== "number") continue;
    out.push({ x: p.x, y: p.y });
  }
  return out;
}

function endpoint(raw: unknown): WorldPoint | null {
  if (!isRecord(raw) || typeof raw.x !== "number" || typeof raw.y !== "number") return null;
  return { x: raw.x, y: raw.y };
}

/** Parse `POST /api/preview/route`; returns null when the body is not a route. */
export function parseRoutePreview(raw: unknown, version: number): RoutePreview | null {
  if (!isRecord(raw) || !Array.isArray(raw.legs)) return null;
  const legs: PreviewLeg[] = [];
  raw.legs.forEach((entry, i) => {
    if (!isRecord(entry)) return;
    const stepId = str(entry.step_id);
    const index = num(entry.index, i);
    let points = posePoints(entry.path);
    if (points.length < 2) {
      const a = endpoint(entry.start);
      const b = endpoint(entry.goal);
      points = a && b ? [a, b] : [];
    }
    if (points.length < 2) return;
    legs.push({
      key: `${stepId}:${index}:${i}`,
      stepId,
      index,
      points,
      planned: entry.planned === true,
      error: str(entry.error),
      lengthM: num(entry.length_m),
    });
  });
  return {
    legs,
    distanceM: num(raw.distance_m),
    estimateS: num(raw.estimate_s),
    speedMps: num(raw.speed_mps, 0.5),
    planned: raw.planned === true,
    note: str(raw.note),
    version,
    at: Date.now(),
  };
}

export function formatDistance(m: number): string {
  if (!Number.isFinite(m)) return "";
  if (m < 10) return `${m.toFixed(1)} m`;
  if (m < 1000) return `${Math.round(m * 10) / 10} m`;
  return `${(m / 1000).toFixed(2)} km`;
}

/** "about 1 min 20 s" — deliberately vague, it is an estimate. */
export function formatEstimate(seconds: number): string {
  if (!Number.isFinite(seconds) || seconds <= 0) return "about 0 s";
  const s = Math.round(seconds);
  if (s < 60) return `about ${s} s`;
  const m = Math.floor(s / 60);
  const rest = s - m * 60;
  if (m < 60) return `about ${m} min${rest ? ` ${rest} s` : ""}`;
  const hh = Math.floor(m / 60);
  const mm = m - hh * 60;
  return `about ${hh} h${mm ? ` ${mm} min` : ""}`;
}

/** "18.4 m · about 1 min 20 s". */
export function routeSummary(p: RoutePreview): string {
  return `${formatDistance(p.distanceM)} · ${formatEstimate(p.estimateS)}`;
}

// ---- dry run ---------------------------------------------------------------------------

export interface DrySample {
  t: number;
  x: number;
  y: number;
  yaw: number;
  /** Step id this sample belongs to ("" between steps). */
  step: string;
}

export interface DryStep {
  id: string;
  name: string;
  type: string;
  t0: number;
  t1: number;
  status: string;
  error: string;
}

export interface DryOutput {
  kind: string;
  /** Everything else the runner recorded, for the detail line. */
  detail: string;
}

export interface DryRunResult {
  ok: boolean;
  status: string;
  error: string;
  durationS: number;
  distanceM: number;
  samples: DrySample[];
  steps: DryStep[];
  notes: string[];
  outputs: DryOutput[];
  truncated: boolean;
  version: number;
}

function outputOf(raw: unknown): DryOutput | null {
  if (!isRecord(raw)) return null;
  const kind = str(raw.kind) || "output";
  const parts: string[] = [];
  for (const [k, v] of Object.entries(raw)) {
    if (k === "kind") continue;
    parts.push(`${k}=${typeof v === "string" ? v : JSON.stringify(v)}`);
  }
  return { kind, detail: parts.join(" · ") };
}

/** Parse `POST /api/preview/dryrun`; returns null when the body is not a dry run. */
export function parseDryRun(raw: unknown, version: number): DryRunResult | null {
  if (!isRecord(raw) || !Array.isArray(raw.samples)) return null;
  const samples: DrySample[] = [];
  for (const s of raw.samples) {
    if (!isRecord(s) || typeof s.x !== "number" || typeof s.y !== "number") continue;
    samples.push({ t: num(s.t), x: s.x, y: s.y, yaw: num(s.yaw_deg), step: str(s.step) });
  }
  samples.sort((a, b) => a.t - b.t);
  const steps: DryStep[] = [];
  for (const s of Array.isArray(raw.steps) ? raw.steps : []) {
    if (!isRecord(s)) continue;
    const id = str(s.id);
    if (!id) continue;
    steps.push({ id, name: str(s.name) || id, type: str(s.type), t0: num(s.t0), t1: num(s.t1), status: str(s.status), error: str(s.error) });
  }
  const outputs: DryOutput[] = [];
  for (const o of Array.isArray(raw.outputs) ? raw.outputs : []) {
    const parsed = outputOf(o);
    if (parsed) outputs.push(parsed);
  }
  const last = samples.length ? samples[samples.length - 1]!.t : 0;
  return {
    ok: raw.ok === true,
    status: str(raw.status) || (raw.ok === true ? "succeeded" : "failed"),
    error: str(raw.error),
    durationS: Math.max(num(raw.duration_s), last),
    distanceM: num(raw.distance_m),
    samples,
    steps,
    notes: (Array.isArray(raw.notes) ? raw.notes : []).map((n) => (typeof n === "string" ? n : JSON.stringify(n))),
    outputs,
    truncated: raw.truncated === true,
    version,
  };
}

export interface GhostPose {
  x: number;
  y: number;
  yaw: number;
  step: string;
}

/** The ghost robot's pose at time `t`, interpolated between samples. */
export function poseAt(result: DryRunResult, t: number): GhostPose | null {
  const s = result.samples;
  if (!s.length) return null;
  if (t <= s[0]!.t) return { x: s[0]!.x, y: s[0]!.y, yaw: s[0]!.yaw, step: s[0]!.step };
  const last = s[s.length - 1]!;
  if (t >= last.t) return { x: last.x, y: last.y, yaw: last.yaw, step: last.step };
  let lo = 0;
  let hi = s.length - 1;
  while (hi - lo > 1) {
    const mid = (lo + hi) >> 1;
    if (s[mid]!.t <= t) lo = mid;
    else hi = mid;
  }
  const a = s[lo]!;
  const b = s[hi]!;
  const span = b.t - a.t;
  const f = span > 0 ? (t - a.t) / span : 0;
  let dyaw = b.yaw - a.yaw;
  while (dyaw > 180) dyaw -= 360;
  while (dyaw < -180) dyaw += 360;
  return { x: a.x + (b.x - a.x) * f, y: a.y + (b.y - a.y) * f, yaw: a.yaw + dyaw * f, step: f < 0.5 ? a.step : b.step };
}

/** The path the ghost has driven up to `t` (the trail behind it). */
export function trailUpTo(result: DryRunResult, t: number): WorldPoint[] {
  const out: WorldPoint[] = [];
  for (const s of result.samples) {
    if (s.t > t) break;
    out.push({ x: s.x, y: s.y });
  }
  const head = poseAt(result, t);
  if (head) out.push({ x: head.x, y: head.y });
  return out;
}

/** Status of every step at time `t`: running now, or how it ended before `t`. */
export function stepStatesAt(result: DryRunResult, t: number): Map<string, { status: string; error: string }> {
  const out = new Map<string, { status: string; error: string }>();
  for (const s of result.steps) {
    if (s.t0 > t) continue;
    if (s.t1 > t) out.set(s.id, { status: "running", error: "" });
    else out.set(s.id, { status: s.status || "succeeded", error: s.error });
  }
  return out;
}
